enhancements/network/dual-stack.md

```diff
@@ -0,0 +1,261 @@
+the `IPv6DualStack` feature gate in `kubelet`, `kube-apiserver`, and
+`kube-controller-manager`. We tried to do this for all clusters in 4.4
+but [ran into bugs] and had to revert it. The [upstream bugs] are not
+yet fixed.
+
+To work around this for early CI, we will patch
+`machine-config-operator`, `cluster-kube-apiserver-operator`, and
+`cluster-kube-controller-manager-operator` to add the feature gate to
+their operands if and only if the `install-config.yaml` defines a
+dual-stack cluster. (This won't allow migrating from single-stack to
+dual-stack but testing that can wait until after the upstream problems
+are fixed and we can enable the feature gate everywhere.)
+
+[ran into bugs]: https://bugzilla.redhat.com/show_bug.cgi?id=1794376
+[upstream bugs]: https://github.com/kubernetes/kubernetes/pull/86895
+
+#### Upstream Kubernetes
+
+We need to continue watching/shepherding upstream [the dual-stack KEP]
+and [dual-stack kubernetes PRs]. In theory, not much further work is
+needed upstream for dual-stack on bare metal. (CloudProvider support
+for IPv6/dual-stack is much poorer.)
+
+There are currently very few dual-stack-specific e2e tests upstream
+and we will probably need to create some more.
+
+[the dual-stack KEP]: https://github.com/kubernetes/enhancements/pulls?q=is%3Apr+is%3Aopen+dual-stack
+[dual-stack kubernetes PRs]: https://github.com/kubernetes/kubernetes/issues?q=is%3Aopen+DualStack
+
+#### OVN-Kubernetes
+
+Work is ongoing to add dual-stack support to OVN-Kubernetes.
+([OVN-Kubernetes Dual-Stack Tracker]). Upstream OVN-Kubernetes uses CI
+based on `kind` (Kubernetes-in-Docker) and we will have a dual-stack
+`kind` environment ([kind PR], [ovn-kube PR]) for testing it against
+dual-stack Kubernetes.
+
+[OVN-Kubernetes Dual-Stack Tracker]: https://github.com/ovn-org/ovn-kubernetes/issues/1142
+[kind PR]: https://github.com/kubernetes-sigs/kind/pull/692
+[ovn-kube PR]: https://github.com/ovn-org/ovn-kubernetes/issues/1248
+
+#### Installer
+
+The installer already allows configuring dual-stack clusters on
+bare metal (and Azure) even though they don't yet work, so no work
+should be needed there until/unless we want to support additional
+cloud platforms.
+
+#### Other OCP Components
+
+We believe that *most* components will not need any new work to
+support dual-stack beyond what was done for single-stack IPv6. In
+particular, most services are now listening for both IPv4 and IPv6
+connections regardless of cluster configuration, and all code that
+deals with IP addresses should be capable of dealing with either IPv4
+or IPv6 addresses.
+
+We are currently doing a survey of single-stack-IPv6-related changes
+that were made in the 4.4 cycle to use as a starting point for
+figuring out where further dual-stack changes may be needed. (One
+known problem is with services that only listen on "localhost" rather
+than on all interfaces; there's no trivial way to listen on both
+`127.0.0.1` and `::1`.)
```

Comment:
Does that also apply to /metrics endpoints in operators? Or can we get by with those using only v4 or v6 in a dual-stack setup?

Reply:
Is this the case where there's a proxy in front of the metadata endpoint, and the proxy hits the metadata endpoint via localhost?  In that case, just v4 or just v6 is fine.  We just have to make sure the proxy and metadata endpoint pick the same one.  If they both hard code an IP address, it's no big deal.  If the client side uses `localhost`, then most systems will resolve that to both the IPv4 and IPv6 addresses, so it should try the other if one fails (client dependent, but a good client will ...).

Reply:
There is exactly one proxy in use - kube-rbac-proxy, so it should be easy to make it do the right thing.